### ISSUE TITLE
RCU stall on `tfw_cache_h2_copy_chunked_body()`

### ISSUE THREAD

**Issue by krizhanovsky** [tempesta-tech/tempesta#1970]
`master` as of c43e7d44711a6445540b267410b95c338ab3170e with config
```
# cat /etc/tempesta/tempesta_fw.conf 
cache 2;
cache_fulfill * *;
cache_methods GET HEAD;
cache_ttl 3600;

listen 443 proto=https;
listen 80;

srv_group default {
        server 192.168.122.1:80;
}

#srv_group auth {
#        server 127.0.0.1:8040;
#}

tls_certificate /etc/tempesta/fullchain.pem;
tls_certificate_key /etc/tempesta/privkey.pem;
req_hdr_add X-Forwarded-Proto "https";
resp_hdr_set Strict-Transport-Security "max-age=31536000; includeSubDomains";

vhost default {
        tls_match_any_server_name;
        resp_hdr_set Cache-Control;
        nonidempotent GET * *;
        nonidempotent HEAD * *;
        nonidempotent POST * *;
        proxy_pass default;
}

#vhost auth {
#        proxy_pass auth;
#}

#vhost wp-auth {
#        req_hdr_add Location "wp-admin";
#        proxy_pass auth;
#}

frang_limits {
        request_rate 200;
        http_method_override_allowed true;
        http_methods post put get;
}

cache_purge;
cache_purge_acl 192.168.122.1 127.0.0.1;

access_log on;

block_action attack reply;
block_action error reply;

#http_chain admin {
#        mark == 23  -> default;
#        -> auth;
#}

http_chain {

        # Set cache_ttl for speciefic locations
        uri == "/blog/*" -> cache_ttl = 1;
        uri == "/knowledge-base/*" -> cache_ttl = 600;

        # Set Security
#        mark == 23 -> default;
#        uri == "*/google/*" -> auth;
#        uri == "*/wp-admin/*" -> admin;
#        uri == "*/wp-login.php*" -> admin;
#        uri == "*/phpmyadmin*" -> admin;
        -> default;
}
```
Tempesta FW is running in a VM and querying the new wordpress website from an outside LXD container.

Running (**100% reproducable**):
```
# /usr/bin/wget -q -O /dev/null http://127.0.0.1/index
```
spits this kernel call trace (not a panic though, but the command hungs):

```
[ 3135.113948] rcu: INFO: rcu_sched detected stalls on CPUs/tasks:
[ 3135.115196] rcu: 	6-....: (4 GPs behind) idle=82a/0/0x1 softirq=1714/1714 fqs=1051 
[ 3135.116557] 	(detected by 13, t=2102 jiffies, g=36937, q=1205)
[ 3135.117637] Sending NMI from CPU 13 to CPUs 6:
[ 3131.373971] NMI backtrace for cpu 6
[ 3131.373971] CPU: 6 PID: 0 Comm: swapper/6 Tainted: G           O      5.10.35-tfw-secperf+ #2
[ 3131.373971] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.15.0-1 04/01/2014
[ 3131.373971] RIP: 0010:tfw_cache_h2_copy_chunked_body+0x131/0x260 [tempesta_fw]
[ 3131.373971] Code: 48 29 c2 48 c7 44 24 50 00 00 00 00 48 c7 44 24 58 00 00 00 00 48 89 44 24 40 48 89 54 24 48 48 85 d2 75 3c 4c 89 d8 48 39 d8 <74> 7c 48 39 e8 75 b5 41 83 c5 01 45 3b 6f 18 0f 83 d3 00 00 00 49
[ 3131.373971] RSP: 0018:ffffafb2402148d8 EFLAGS: 00000283
[ 3131.373971] RAX: ffff9491a7735b3f RBX: ffff9491d05ee7c0 RCX: 0000000000000000
[ 3131.373971] RDX: ffff9491a2338ed0 RSI: ffffafb2402149d0 RDI: ffff94918f229038
[ 3131.373971] RBP: ffff9491a7737eb3 R08: ffff9491a23381c0 R09: ffffafb2402149c0
[ 3131.373971] R10: ffff9491a23381c0 R11: ffff9491d05ee600 R12: ffff9491d05ee782
[ 3131.373971] R13: 0000000000000001 R14: 0000000000000002 R15: ffff9491a23381c0
[ 3131.373971] FS:  0000000000000000(0000) GS:ffff9494ef980000(0000) knlGS:0000000000000000
[ 3131.373971] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 3131.373971] CR2: 00007f1150b55e24 CR3: 000000003260a006 CR4: 0000000000770ee0
[ 3131.373971] PKRU: 55555554
[ 3131.373971] Call Trace:
[ 3131.373971]  <IRQ>
[ 3131.373971]  tfw_cache_copy_resp.isra.0+0xa5f/0xbe0 [tempesta_fw]
[ 3131.373971]  __cache_add_node+0xbc/0xd0 [tempesta_fw]
[ 3131.373971]  ? tfw_h2_resp_adjust_fwd+0x4c0/0x4c0 [tempesta_fw]
[ 3131.373971]  tfw_cache_add+0x17c/0x1b0 [tempesta_fw]
[ 3131.373971]  tfw_cache_process+0xa4/0x250 [tempesta_fw]
[ 3131.373971]  ? tfw_gfsm_move+0x18f/0x220 [tempesta_fw]
[ 3131.373971]  tfw_http_resp_cache+0xe6/0x180 [tempesta_fw]
[ 3131.373971]  tfw_http_resp_process+0x34f/0x420 [tempesta_fw]
[ 3131.373971]  ? __dev_queue_xmit+0x2bb/0x480
[ 3131.373971]  tfw_http_msg_process_generic+0x7f/0x1f0 [tempesta_fw]
[ 3131.373971]  ? ip_finish_output2+0x18a/0x510
[ 3131.373971]  tfw_connection_recv+0x70/0xe0 [tempesta_fw]
[ 3131.373971]  ss_tcp_process_skb.constprop.0+0x124/0x270 [tempesta_fw]
[ 3131.373971]  ss_tcp_process_data+0xdc/0x170 [tempesta_fw]
[ 3131.373971]  ss_tcp_data_ready+0x3a/0xc0 [tempesta_fw]
[ 3131.373971]  tcp_rcv_established+0x57d/0x620
[ 3131.373971]  tcp_v4_do_rcv+0x126/0x1e0
[ 3131.373971]  tcp_v4_rcv+0xbb4/0xce0
[ 3131.373971]  ? tcp_v4_early_demux+0xa3/0x140
[ 3131.373971]  ? tcp_v4_early_demux+0xe7/0x140
[ 3131.373971]  ip_protocol_deliver_rcu+0x14/0x170
[ 3131.373971]  ip_local_deliver_finish+0x3f/0x50
[ 3131.373971]  __netif_receive_skb_one_core+0x40/0x50
[ 3131.373971]  process_backlog+0x83/0x120
[ 3131.373971]  napi_poll+0x8a/0x1b0
[ 3131.373971]  net_rx_action+0x95/0x180
[ 3131.373971]  __do_softirq+0xc0/0x26e
[ 3131.373971]  asm_call_irq_on_stack+0xf/0x20
[ 3131.373971]  </IRQ>
[ 3131.373971]  do_softirq_own_stack+0x32/0x40
[ 3131.373971]  irq_exit_rcu+0x83/0xb0
[ 3131.373971]  sysvec_call_function_single+0x2e/0x80
[ 3131.373971]  asm_sysvec_call_function_single+0x12/0x20
[ 3131.373971] RIP: 0010:default_idle+0xe/0x10
[ 3131.373971] Code: fb 65 48 8b 04 25 00 6d 01 00 f0 80 60 02 df c3 0f ae f0 0f ae 38 0f ae f0 eb b6 90 e9 07 00 00 00 0f 00 2d 16 c8 47 00 fb f4 <c3> cc 65 8b 15 c9 89 a8 59 89 d2 48 8b 05 30 26 67 01 48 03 04 d5
[ 3131.373971] RSP: 0018:ffffafb24009bef0 EFLAGS: 00000206
[ 3131.373971] RAX: ffff9494ef9aab40 RBX: ffff9491a0230000 RCX: ffff9494ef9aab40
[ 3131.373971] RDX: 000000000000e826 RSI: 0000000000000006 RDI: ffff9494ef99e180
[ 3131.373971] RBP: 0000000000000000 R08: 000000007c5145a0 R09: 0000000000000040
[ 3131.373971] R10: 0000000000000001 R11: 0000000000000000 R12: 0000000000000000
[ 3131.373971] R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
[ 3131.373971]  default_idle_call+0x2d/0xa0
[ 3131.373971]  cpuidle_idle_call+0x133/0x190
[ 3131.373971]  ? tsc_verify_tsc_adjust+0x2d/0xc0
[ 3131.373971]  do_idle+0x68/0xb0
[ 3131.373971]  cpu_startup_entry+0x14/0x20
[ 3131.373971]  secondary_startup_64_no_verify+0xb0/0xbb
```

**Comment by krizhanovsky** [comment]
```
# addr2line -e tempesta_fw.ko -i $(perl -e 'printf("%x\n", 0x'$(nm tempesta_fw.ko|grep '\<tfw_cache_h2_copy_chunked_body$'|awk '{print $1}')'+0xf0)')
/root/tempesta/fw/cache.c:1309

# cat -n cache.c +1309 | head -1320 | tail -20
  1301				if (++c_chunk >= cut->nchunks)
  1302					return 0;
  1303				curr += stop_len;
  1304				tmp = __TFW_STR_CH(cut, c_chunk);
  1305				stop = tmp->data;
  1306				stop_len = tmp->len;
  1307			}
  1308	
  1309			if ((begin <= curr) && (end > curr)) {
  1310				TfwStr chunk = {.len = 0};
  1311	
  1312				/* Stop found in current frag. */
  1313				if ((begin <= stop) && (end >= stop)) {
  1314					chunk.data = curr;
  1315					chunk.len = stop - curr;
  1316					curr = stop;
  1317				} else {
  1318					/* Stop not found. Copy whole frag. */
  1319					chunk.data = curr;
  1320					chunk.len = end - curr;
```

**Comment by krizhanovsky** [comment]
The bug is crucial since it doesn't allow a local keepalive instance to check the status of Tempesta FW.

**Comment by krizhanovsky** [comment]
Instrumentation patch
```diff
diff --git a/fw/cache.c b/fw/cache.c
index 20f03a14..d28f34d7 100644
--- a/fw/cache.c
+++ b/fw/cache.c
@@ -1282,6 +1282,7 @@ tfw_cache_h2_copy_chunked_body(unsigned int *acc_len, char **p, TdbVRec **trec,
        stop_len = tmp->len;
 
        while (true) {
+pr_err("1970-DBG %d: cut.nchunks=%u it.frag=%d nr_frags=%u headlen=%u it.skb=%p stop=%p curr=%p prev_end=%p\n", __LINE__, cut->nchunks, it.frag, skb_shinfo(it.skb)->nr_frags, skb_headlen(it.skb), it.skb, stop, curr, prev_end);
                if (it.frag == -1) {
                        begin = it.skb->data;
                        end = begin + skb_headlen(it.skb);
@@ -1291,11 +1292,13 @@ tfw_cache_h2_copy_chunked_body(unsigned int *acc_len, char **p, TdbVRec **trec,
                        begin = skb_frag_address(f);
                        end = begin + skb_frag_size(f);
                }
+pr_err("1970-DBG %d: begin=%p end=%p\n", __LINE__, begin, end);
 
                /* Previous fragment is fully copied. */
                if (curr == prev_end)
                        curr = begin;
 rep:
+pr_err("1970-DBG %d: stop=%p stop_len=%d curr=%p c_chunk=%d\n", __LINE__, stop, stop_len, curr, c_chunk);
                /* Stop is found. Get next stop. */
                if (curr == stop) {
                        if (++c_chunk >= cut->nchunks)
@@ -1306,6 +1309,7 @@ rep:
                        stop_len = tmp->len;
                }
 
+pr_err("1970-DBG %d: begin=%p end=%p stop=%p stop_len=%d curr=%p\n", __LINE__, begin, end, stop, stop_len, curr);
                if ((begin <= curr) && (end > curr)) {
                        TfwStr chunk = {.len = 0};
 
```
produces
```
[   58.291142] 1970-DBG 1285: cut.nchunks=4 it.frag=-1 nr_frags=1 headlen=62 it.skb=00000000ee58626d stop=00000000570117c5 curr=00000000570117c5 prev_end=0000000000000000
[   58.293725] 1970-DBG 1295: begin=00000000f549792a end=000000006ad2cc24
[   58.295026] 1970-DBG 1301: stop=00000000570117c5 stop_len=6 curr=00000000570117c5 c_chunk=0
[   58.296531] 1970-DBG 1312: begin=00000000f549792a end=000000006ad2cc24 stop=00000000d8688e5d stop_len=2 curr=00000000d06512de
[   58.298494] 1970-DBG 1301: stop=00000000d8688e5d stop_len=2 curr=00000000d06512de c_chunk=1
[   58.300004] 1970-DBG 1312: begin=00000000f549792a end=000000006ad2cc24 stop=00000000d8688e5d stop_len=2 curr=00000000d06512de
[   58.301595] 1970-DBG 1301: stop=00000000d8688e5d stop_len=2 curr=00000000d06512de c_chunk=1
[   58.302169] 1970-DBG 1312: begin=00000000f549792a end=000000006ad2cc24 stop=00000000d8688e5d stop_len=2 curr=00000000d06512de
[   58.303662] 1970-DBG 1301: stop=00000000d8688e5d stop_len=2 curr=00000000d06512de c_chunk=1
[   58.305625] 1970-DBG 1312: begin=00000000f549792a end=000000006ad2cc24 stop=00000000d8688e5d stop_len=2 curr=00000000d06512de
[   58.307134] 1970-DBG 1301: stop=00000000d8688e5d stop_len=2 curr=00000000d06512de c_chunk=1
[   58.309015] 1970-DBG 1312: begin=00000000f549792a end=000000006ad2cc24 stop=00000000d8688e5d stop_len=2 curr=00000000d06512de
....
```

**Comment by krizhanovsky** [comment]
The dump of problem data:
```
[59799.816182] [tempesta fw] SKB (ffff9ba96ce01b00) DUMP: len=7240 data_len=7178 truesize=8448 users=1^M
[59799.817324] [tempesta fw]   head=ffff9ba96ce01c00 data=ffff9ba96ce01c82 tail=c0 end=c0^M
[59799.818375] [tempesta fw]   nr_frags=1 frag_list=0000000000000000 next=ffff9ba96ce01300 prev=ffff9ba960a8b300^M
[59799.819995] [tempesta fw]   head data (62):^M
[59799.820891]     00000000: 48 54 54 50 2f 31 2e 31 20 32 30 30 20 4f 4b 0d  HTTP/1.1 200 OK.^M
[59799.821887]     00000010: 0a 44 61 74 65 3a 20 57 65 64 2c 20 30 36 20 53  .Date: Wed, 06 S^M
[59799.822996]     00000020: 65 70 20 32 30 32 33 20 31 32 3a 31 30 3a 33 30  ep 2023 12:10:30^M
[59799.824566]     00000030: 20 47 4d 54 0d 0a 53 65 72 76 65 72 3a 20         GMT..Server: ^M
[59799.825635] [tempesta fw]   frag  0 (addr=ffff9ba96c561e8c pg_off=7820 size=7178 pg_ref=4):^M
[59799.826130]     00000000: 41 70 61 63 68 65 2f 32 2e 34 2e 35 32 20 28 55  Apache/2.4.52 (U^M
[59799.827179]     00000010: 62 75 6e 74 75 29 0d 0a 56 61 72 79 3a 20 41 63  buntu)..Vary: Ac^M
[59799.828318]     00000020: 63 65 70 74 2d 45 6e 63 6f 64 69 6e 67 0d 0a 4b  cept-Encoding..K^M
[59799.829498]     00000030: 65 65 70 2d 41 6c 69 76 65 3a 20 74 69 6d 65 6f  eep-Alive: timeo^M
[59799.830837]     00000040: 75 74 3d 35 2c 20 6d 61 78 3d 31 30 30 0d 0a 43  ut=5, max=100..C^M
[59799.832327]     00000050: 6f 6e 6e 65 63 74 69 6f 6e 3a 20 4b 65 65 70 2d  onnection: Keep-^M
[59799.834166]     00000060: 41 6c 69 76 65 0d 0a 54 72 61 6e 73 66 65 72 2d  Alive..Transfer-^M
[59799.834957]     00000070: 45 6e 63 6f 64 69 6e 67 3a 20 63 68 75 6e 6b 65  Encoding: chunke^M
[59799.836476]     00000080: 64 0d 0a 43 6f 6e 74 65 6e 74 2d 54 79 70 65 3a  d..Content-Type:^M
[59799.837651]     00000090: 20 74 65 78 74 2f 68 74 6d 6c 3b 20 63 68 61 72   text/html; char^M
[59799.839002]     000000a0: 73 65 74 3d 55 54 46 2d 38 0d 0a 0d 0a 32 31 62  set=UTF-8....21b^M             <== chunk length descriptor (4 bytes plus CRLF, 6 in total)
[59799.840504]     000000b0: 63 0d 0a 69 6d 70 6f 72 74 20 68 74 6d 6c 0a 69  c..import html.i^M
[59799.841595]     000000c0: 6d 70 6f 72 74 20 6d 61 72 6b 64 6f 77 6e 0a 69  mport markdown.i^M
[59799.844158]     000000d0: 6d 70 6f 72 74 20 6f 73 0a 69 6d 70 6f 72 74 20  mport os.import ^M
...
```
with the first iteration
```
[59800.217803] 1970-DBG 1286: cut.nchunks=4 itfrag=-1 nr_frags=1 headlen=62 it.skb=ffff9ba96ce01b00 stop=ffff9ba96c561f39 curr=ffff9ba96c561f39 prev_end=0000000000000000^M
[59800.218631] 1970-DBG 1297: begin=ffff9ba96ce01c82 end=ffff9ba96ce01cc0 headlen=62^M
[59800.220021] 1970-DBG 1307: stop=ffff9ba96c561f39 stop_len=6 curr=ffff9ba96c561f39 c_chunk=0^M
```

**Comment by krizhanovsky** [comment]
Fixed in #1972
